### PR TITLE
[RTM] Use Symfony asset component for Contao assets

### PR DESCRIPTION
**Pull request by aschempp**
I started using Symfony Encore for our application asset management and stumbled over the Symfony Asset service. I think there are a ton of useful feature for Contao.

 - It *knows* the path to an asset by asking for the composer package.
 - It automatically adds the `TL_ASSETS_URL` prefix if configured.
 - It automatically adds the package version to the URL, which results in clever cache busting on updates.

The provided code is totally working, but there are a lot more features we can implement. Missing **TODOs**:
 - [ ] Use asset URLs for `TL_JAVASCRIPT` and `TL_CSS`
 - [ ] Add support for asset URLs (absolute) to the `Contao\Combiner`
 - [x] Re-implement `Controller::setStaticUrls` instead of using constants in the context service.
 - [x] *Maybe* add `Resources/public` of all bundles as packages.
 - [ ] *Maybe* add a package for the backend theme
 - [x] Unit Tests

### CODE REVIEW

File: src/Asset/ContaoContext.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace Contao\CoreBundle\Asset;
+
+use Symfony\Component\Asset\Context\ContextInterface;
+
+class ContaoContext implements ContextInterface
```
Comment (Toflar) [inline src/Asset/ContaoContext.php]: Nothing should rely on the constants anymore. If you need to access `TL_*` it's better to find a different way because the constants will not be available in 5 anymore.

Comment (leofeyer) [reply src/Asset/ContaoContext.php]: The only other way is to read from `$GLOBALS` in the back end and from `global $objPage` in the front end, which does not seem less wrong than using the constants to me.

File: src/Asset/ContaoContext.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * Gets the base path.
+     *
+     * @return string The base path
+     */
+    public function getBasePath()
+    {
+        if (!defined($this->constant)) {
+            return '';
+        }
```
Comment (DanielSchwiperich) [inline src/Asset/ContaoContext.php]: is this check needed? `constant` cannot be undefined due to constructor string requirement and no method which unsets the property

File: src/Asset/ContaoContext.php (hunk 1 above)
Comment (Toflar) [reply src/Asset/ContaoContext.php]: Yeah, or we find a new way because apparently we will still need that feature in the future :-)

Comment (aschempp) [reply src/Asset/ContaoContext.php]: Added ToDo no. 3 for this. We should probably inject the PageModel and "duplicate" the code from `Controller::setStaticUrls`.

**Comment by aschempp** [comment]
I've added a dependency to https://github.com/terminal42/asset-bundle. This will automatically add all public bundle files as asset packages, completing the new Contao setup.


**in Twig templates:**
```twig
<img src="{{ asset('img/icon.png', 'rock_solid_slider') }}">
<img src="{{ asset('folder.gif', 'terminal42_folderpage') }}">
```

**in Contao templates:**
```html
<img src="<?= $this->asset('img/icon.png', 'rock_solid_slider') ?>">
<img src="<?= $this->asset('folder.gif', 'terminal42_folderpage') ?>">
```

**in Contao HTML module/element:**
```html
<img src="{{asset::img/icon.png::rock_solid_slider}}">
<img src="{{asset::folder.gif::terminal42_folderpage}}">
```

**Comment by leofeyer** [comment]
I think we should finish this PR with the current feature set. The only item from the todo list that needs to be done is "Unit Tests". The rest can be added later.

**Comment by aschempp** [comment]
@Toflar I have rewritten the contexts to use the page model, is that how you intended it?

File: src/Asset/ContaoContext.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        return (bool) $this->page->loadDetails()->rootUseSSL;
+    }
+
+    /**
+     * @return string
+     */
+    private function getFieldValue(): string
+    {
+        if (null === $this->page) {
```
Comment (leofeyer) [inline src/Asset/ContaoContext.php]:
```php
        if (null !== $this->page) {
            return (string) $this->page->{$this->field};
        }

        /** @var Config $config */
        $config = $this->framework->createInstance(Config::class);

        return (string) $config->get($this->field);
```

**Comment by aschempp** [comment]
> Why not have a ContaoAssetsContext and ContaoFilesContext?

I don't think that would make anything better. It means we need three classes instead of one, including an abstract. Which will be exactly the same, a class that will be configured, just by extending instead of a constructor argument.

File: src/Asset/ContaoContext.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+     *
+     * @param PageModel|null $page
+     *
+     * @return string
+     */
+    private function getFieldValue(?PageModel $page): string
+    {
+        if (null === $page) {
+            /** @var Config $config */
+            $config = $this->framework->createInstance(Config::class);
```
Comment (leofeyer) [inline src/Asset/ContaoContext.php]: Why are you using `createInstance()` instead of `getAdapter()` here? And shouldn't we initialize the framework first?

File: src/Resources/contao/library/Contao/Model.php [hunk 5]
```diff
@@ -394,7 +394,7 @@ public function mergeRow(array $arrData)
 	 */
 	public function markModified($strKey)
 	{
-		if (!isset($this->arrModified[$strKey]))
+		if (!isset($this->arrModified[$strKey]) && isset($this->arrData[$strKey]))
```
Comment (leofeyer) [inline src/Resources/contao/library/Contao/Model.php]: Is this change related to the feature?

File: src/Asset/ContaoContext.php (hunk 4 above)
Comment (Toflar) [reply src/Asset/ContaoContext.php]: Both must be fixed indeed.

File: src/Resources/contao/library/Contao/Controller.php [hunk 6]
```diff
@@ -1792,45 +1793,31 @@ public static function addEnclosuresToTemplate($objTemplate, $arrItem, $strKey='
 	/**
 	 * Set the static URL constants
 	 *
-	 * @param PageModel $objPage An optional page object
+	 * @deprecated Deprecated in Contao 4.5, to be removed in Contao 5. Use the asset contexts instead.
```
Comment (leofeyer) [inline src/Resources/contao/library/Contao/Controller.php]: Since we have deprecated the method, we should no longer use it. Otherwise, if we still need to use it, we should not deprecate it. 😉

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Comment by fritzmg** [comment]
In your examples you are using
```php
<img src="<?= $this->asset('folder.gif', 'terminal42_folderpage') ?>">
```
however, shouldn't it be
```php
<img src="<?= $this->asset('folder.gif', 'terminal42/contao-folderpage') ?>">
```
Or what exactly is `$packageName`?

**Comment by Toflar** [comment]
Yes it is, the example comment was before some more commits ;)

**Comment by aschempp** [comment]
@fritzmg `terminal42_folder` is correct. The bundle assets are based on the bundle classname, which in most cases and for correct setups should match the bundle alias for container configuration. We can't use the package name because we don't know which package a bundle belongs to (and a package theoretically can contain multiple bundles).

The diffeference is Contao componentes (assets), where the package name is the composer package (e.g. `contao-components/jquery`), because these do not have a bundle name.
